libexec/rtld-elf/rtld_c18n.c

```diff
@@ -0,0 +1,700 @@
+} tramp_pgs = {
+	.lock = ATOMIC_FLAG_INIT
+};
+
+static slot_idx_t
+tramp_table_max_load(int exp)
+{
+	/* MAX_LOAD is 37.5% of capacity. */
+	return (3 << (exp - 3));
+}
+
+static struct tramp_map_kv *
+tramp_map_new(int exp)
+{
+	struct tramp_map_kv *map;
+
+	map = xmalloc(sizeof(*map) << exp);
+	for (size_t i = 0; i < (1 << exp); ++i)
+		map[i] = (struct tramp_map_kv) {
+			.key = 0,
+			.index = -1
+		};
+	return (map);
+}
+
+static struct tramp_data *
+tramp_data_expand(int exp)
+{
+	struct tramp_data *data;
+	/*
+	 * The table only needs to be as large as the MAX_LOAD.
+	 */
+	data = realloc(tramp_table.data, sizeof(*data) * tramp_table_max_load(exp));
+	if (data == NULL)
+		rtld_fatal("realloc failed");
+	return (data);
+}
+
+/* Taken from https://github.com/skeeto/hash-prospector/issues/19 */
+static uint32_t
+pointerHash(uint64_t key) {
+	uint32_t x = key ^ (key >> 32);
+	x ^= x >> 16;
+	x *= 0x21f0aaadU;
+	x ^= x >> 15;
+	x *= 0xd35a2d97U;
+	x ^= x >> 15;
+	return (x);
+}
+
+static slot_idx_t
+nextSlot(slot_idx_t slot, uint32_t hash, int exp)
+{
+	uint32_t mask = (1 << exp) - 1;
+	uint32_t step = (hash >> (32 - exp)) | 1;
+	return ((slot + step) & mask);
+}
+
+static void
+resizeTable(int exp)
+{
+	struct tramp_map_kv *map;
+	struct tramp_data *data;
+	ptraddr_t key;
+	uint32_t hash;
+	slot_idx_t size, slot;
+
+	assert(0 < exp && exp < 32);
+
+	free(tramp_table.map);
+	data = tramp_data_expand(exp);
+	map = tramp_map_new(exp);
+	size = atomic_load_explicit(&tramp_table.size, memory_order_relaxed);
+
+	for (slot_idx_t idx = 0; idx < size; ++idx) {
+		key = (ptraddr_t)data[idx].target;
+		hash = pointerHash(key);
+		slot = hash;
+		do
+			slot = nextSlot(slot, hash, exp);
+		while (atomic_load_explicit(
+		    &map[slot].key, memory_order_relaxed) != 0);
+		atomic_store_explicit(&map[slot].key, key,
+		    memory_order_relaxed);
+		atomic_store_explicit(&map[slot].index, idx,
+		    memory_order_relaxed);
+	}
+
+	tramp_table.data = data;
+	tramp_table.map = map;
+	tramp_table.exp = exp;
+}
+
+void
+tramp_hook(int event, void *target, const Obj_Entry *obj, const Elf_Sym *def,
+    void *link, void *rcsp)
+{
+	Elf64_Word sym_num;
+	const char *sym;
+	const char *callee;
+
+	compart_id_t caller_id;
+	const char *caller;
+
+	sym_num = def == NULL ? 0 : def->st_name;
+	sym = def == NULL ? "<unknown>" : strtab_value(obj, def->st_name);
+	callee = comparts.data[obj->compart_id]->name;
+
+	if ((cheri_getperm(link) & CHERI_PERM_EXECUTIVE) == 0)
+		caller_id = ((uintptr_t *)
+		    cheri_setaddress(rcsp, cheri_gettop(rcsp)))[-2];
+	else
+		caller_id = 0;
+	caller = comparts.data[caller_id]->name;
+
+	if (ld_compartment_utrace != NULL)
+		ld_utrace_log(event,
+			target, NULL, sym_num, 0, callee, sym, caller);
+	if (ld_compartment_overhead != NULL)
+		getpid();
+}
+
+static void *
+tramp_pgs_append(const struct tramp_data *data)
+{
+	size_t len;
+	/* A capability-aligned buffer large enough to hold a trampoline */
+	tramp tmp[48];
```

Comment:
48 is too much of a magic constant

Reply:
What would you suggest? A global buffer?

Reply:
#define somewhere?